modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+
+	logger log.Logger
+
+	name           string
+	tenantID       string
+	workerCount    int
+	size           int
+	reqChan        chan *request[T]
+	fn             ProcessFunc[T]
+	workersCloseCh chan struct{}
+
+	pushesTotalMetrics         *prometheus.CounterVec
+	pushesFailuresTotalMetrics *prometheus.CounterVec
+	lengthMetric               *prometheus.GaugeVec
+
+	readOnly *atomic.Bool
+}
+
+func New[T any](cfg Config, logger log.Logger, reg prometheus.Registerer, fn ProcessFunc[T]) *Queue[T] {
+	pushesTotalMetrics := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_total",
+		Help:      "Total number of successful requests queued up for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	pushesFailuresTotalMetric := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	lengthMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"name", "tenant"})
+
+	var alreadyRegisteredErr prometheus.AlreadyRegisteredError
+	err := reg.Register(pushesTotalMetrics)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_total metric", "err", err)
+	}
+	err = reg.Register(pushesFailuresTotalMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_failures_total metric", "err", err)
+	}
+
+	err = reg.Register(lengthMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_length metric", "err", err)
+	}
+
+	return &Queue[T]{
+		logger:                     logger,
+		name:                       cfg.Name,
+		tenantID:                   cfg.TenantID,
+		workerCount:                cfg.WorkerCount,
+		size:                       cfg.Size,
+		reqChan:                    make(chan *request[T], cfg.Size),
+		fn:                         fn,
+		workersCloseCh:             make(chan struct{}),
+		pushesTotalMetrics:         pushesTotalMetrics,
+		pushesFailuresTotalMetrics: pushesFailuresTotalMetric,
+		lengthMetric:               lengthMetric,
+		readOnly:                   atomic.NewBool(false),
+	}
+}
+
+// Push pushes data onto a queue.
+// If the queue is full, the data is dropped
+func (m *Queue[T]) Push(ctx context.Context, data T) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	m.pushesTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	req := &request[T]{
+		data: data,
+	}
+
+	select {
+	case <-ctx.Done():
```

Comment:
why did we split this select? if the chan is blocked and context is cancelled this will no longer exit immediately.

Reply:
yup. agree with this change. i misunderstood what was happening here. thx for the clear explanation!